Rangeselector tests

RosarioPulella opened the pull request:
<!-- 🚨 Please Do Not skip any instructions and information mentioned below as they are all required and essential to evaluate and test the PR. By fulfilling all the required information you will be able to reduce the volume of questions and most likely help merge the PR faster 🚨 -->

<!-- 📝 It is preferred if you keep the "☑️ Allow edits by maintainers" checked in the Pull Request Template as it increases collaboration with the Toolkit maintainers by permitting commits to your PR branch (only) created from your fork.  This can let us quickly make fixes for minor typos or forgotten StyleCop issues during review without needing to wait on you doing extra work. Let us help you help us! 🎉 -->


## Fixes Adds tests for the range selector properties.
<!-- Add the relevant issue number after the "#" mentioned above (for ex: Fixes #1234) which will automatically close the issue once the PR is merged. -->

<!-- Add a brief overview here of the feature/bug & fix. -->
Added unit tests for code interactions with the range selector properties.

## PR Type
What kind of change does this PR introduce?
<!-- Please uncomment one or more that apply to this PR. -->

<!-- - Bugfix -->
<!-- - Feature -->
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Sample app changes -->
<!-- - Other... Please describe: -->


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
No tests for the RangeSelector properties.

## What is the new behavior?
<!-- Describe how was this issue resolved or changed? -->
Tests for the RangeSelector properties.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: <!-- docs PR link -->
- [ ] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)
- [ ] New major technical changes in the toolkit have or will be added to the [Wiki](https://github.com/windows-toolkit/WindowsCommunityToolkit/wiki) e.g. build changes, source generators, testing infrastructure, sample creation changes, etc...
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [ ] Contains **NO** breaking changes

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected within minor release cycles or held until major versions. -->


## Other information

RosarioPulella reviewed (commented):
@azchohfi I added this code to be able to use this syntax for defining a record for these tests. From what I read it seem like its not a big hack or anything, do you think its fine to add? It should only effect our testing code.

RosarioPulella commented:
@azchohfi @Sergio0694 @michael-hawker @shweaver-MSFT @deltakosh @skendrot I am still looking to expand these tests, but I would like some feedback. Any thoughts about what behavior this shows in the `RangeSelector` or if the test code would be greatly appreciated.

azchohfi commented:
I loved that you used the equality of records to Assert results, but I'm not sure about the benefits with this solution, versus the complexity. I'm assuming you did this just to make sure that the test is asserting only one thing, but I would be fine to assert each property individually just for the simplicity. Not sure what others think. I'm not saying I'm not ok with this implementation (and its workaround for records in dot net native), but I'm not sure if it is really a benefit here. You have the parameter values in the same method, and you could just assert them individually. I'm fine with either.

RosarioPulella commented:
> I loved that you used the equality of records to Assert results, but I'm not sure about the benefits with this solution, versus the complexity. I'm assuming you did this just to make sure that the test is asserting only one thing, but I would be fine to assert each property individually just for the simplicity.

I want to assert on the equality of all the properties together as a record so that if any of the values are not as expected then instead of seeing the just one value that is not expected we get to see the values of all the properties, giving use more information in CI without us having to expand the test manually or set break points. For this control its generally important because for many of the properties if you change one it could effect others.

azchohfi commented:
IMO, that is enough explanation to keep it this way then! Makes sense!

michael-hawker commented:
I like the idea that a failure provides all the info about the case that's failing compared to a single value too.

Let's snap this in if we think they're good, then work towards adding more cases or adjusting cases to what we want the desired behavior to be (especially around the snap points). Then we can use that to make modifications to the existing logic (as we know they're broken cases).

chingucoding commented on UnitTests/UnitTests.UWP/UI/Controls/Test_RangeSelector.cs:
```diff
@@ -0,0 +1,343 @@
+
+        // Given:Sf   Min   Max   Start End      Set:Minimum       Then:Sf  Min     Max     Start   End
+#pragma warning disable SA1008, SA1025
+        [DataRow(   1,    0,   10,   90,  100,             0,             1,    0   ,  10   ,  90   , 100   )]
+        [DataRow(   1,    0,   10,   90,  100,           -10,             1,  -10   ,  10   ,  90   , 100   )]
+        [DataRow(   1,    0,   10,   90,  100,            10,             1,   10   ,  10   ,  90   , 100   )]
+        [DataRow(   1,    0,   10,   90,  100,            50,             1,   50   ,  50   ,  90   , 100   )]
+        [DataRow(   1,    0,   10,   90,  100,            90,             1,   90   ,  90   ,  90   , 100   )]
+        [DataRow(   1,    0,   10,   90,  100,           100,             1,  100   , 100   , 100   , 100   )]
+        [DataRow(   1,    0,   10,   90,  100,           110,             1,  110   , 110.01, 110.01, 110.01)]
```
If the first 5 values are always the same, is there a point in providing them everytime or would it be better to scratch them as parameter for easier readability?

RosarioPulella replied:
Good point, took that idea and removed a lot of the repetition from the other tests also, by splitting them up more. I think its more readable overall, lmk what you think.

chingucoding replied:
I think it really helps a lot, it's easier to read why those testcases are there since you can focus on the values instead of reading multiple values that are not actually important to the test.

chingucoding commented on UnitTests/UnitTests.UWP/IsExternalInit.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace System.Runtime.CompilerServices
+{
+    internal static class IsExternalInit
```
Why is this here? According to [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.isexternalinit?view=net-5.0) we shouldn't use that name should we?🤔

chingucoding replied:
Is it worth adding a comment about this then? Or do we expect people to know intuitively why this is here/why it is not a bad idea?

RosarioPulella replied:
I'll add a comment. It would be more clear than the having something in the commit message. Thanks @chingucoding!

RosarioPulella replied:
I added a bit of an explanation and some useful links on the topic.

chingucoding commented on UnitTests/UnitTests.UWP/UI/Controls/Test_RangeSelector.cs:
```diff
@@ -0,0 +1,365 @@
+        // RangeEnd will be Maximum
+
+        // Input:Start End    Expected:Min   Start End   Max
+        [DataRow(    0,    0,              0,    0,    0, 0.01, DisplayName = "Minimum == RangeStart == RangeEnd == Maximum")]
+        [DataRow(    0,   10,              0,    0, 0.01, 0.01, DisplayName = "Minimum == RangeStart  < RangeEnd  > Maximum")]
+        [DataRow(    0,  -10,              0,    0,    0, 0.01, DisplayName = "Minimum == RangeStart  > RangeEnd  < Maximum")]
+        [DataRow(  -10,    0,              0,    0,    0, 0.01, DisplayName = "Minimum  > RangeStart  < RangeEnd == Maximum")]
+        [DataRow(   10,    0,              0,    0,    0, 0.01, DisplayName = "Minimum  < RangeStart  > RangeEnd == Maximum")] 
+        [DataRow(   10,   90,              0, 0.01, 0.01, 0.01, DisplayName = "Minimum  < RangeStart  < RangeEnd  > Maximum")]
+        [DataRow(   90,   10,              0, 0.01, 0.01, 0.01, DisplayName = "Minimum  < RangeStart  > RangeEnd  > Maximum")]
+        [DataRow(  -90,  -10,              0,    0,    0, 0.01, DisplayName = "Minimum  > RangeStart  < RangeEnd  < Maximum")]
+        [DataRow(  -10,  -90,              0,    0,    0, 0.01, DisplayName = "Minimum  > RangeStart  > RangeEnd  < Maximum")]
+        [DataRow(   10,  -10,              0,    0,    0, 0.01, DisplayName = "Minimum  < RangeStart  > RangeEnd  < Maximum")]
+        [DataRow(  -10,   10,              0,    0, 0.01, 0.01, DisplayName = "Minimum  > RangeStart  < RangeEnd  > Maximum")]
+        public Task Initialize_MinEqMax(double rangeStart, double rangeEnd, double expectedMinimum, double expectedRangeStart, double expectedRangeEnd, double expectedMaximum)
+            => Initialize(1, 0, rangeStart, rangeEnd, 0, 1, expectedMinimum, expectedRangeStart, expectedRangeEnd, expectedMaximum);
+
+        [TestCategory("Initialize")]
+        [TestMethod]
+        [DataRow(    0,  100,              0,    0, 0.01, 0.01, DisplayName = "Minimum  > RangeStart  < RangeEnd  > Maximum, RangeStart == Maximum, RangeEnd == Minimum")]
+        [DataRow(   10,   90,              0, 0.01, 0.01, 0.01, DisplayName = "Minimum  > RangeStart  < RangeEnd  > Maximum")]
+        [DataRow(   50,   50,              0, 0.01, 0.01, 0.01, DisplayName = "Minimum  > RangeStart == RangeEnd  > Maximum")]
+        [DataRow(    0,    0,              0,    0,    0, 0.01, DisplayName = "Minimum  > RangeStart == RangeEnd == Maximum")]
+        [DataRow(  100,  100,              0, 0.01, 0.01, 0.01, DisplayName = "Minimum == RangeStart == RangeEnd  > Maximum")]
+
+        [DataRow(   10,  100,              0, 0.01, 0.01, 0.01, DisplayName = "Minimum  > RangeStart  < RangeEnd  > Maximum, RangeEnd   == Maximum")]
+
+        [DataRow(   90,   10,              0, 0.01, 0.01, 0.01, DisplayName = "Minimum  > RangeStart  > RangeEnd  > Maximum")]
+        [DataRow(  100,   10,              0, 0.01, 0.01, 0.01, DisplayName = "Minimum  > RangeStart  > RangeEnd  > Maximum, RangeStart == Maximum")]
+        [DataRow(   90,    0,              0,    0,    0, 0.01, DisplayName = "Minimum  > RangeStart  > RangeEnd  > Maximum, RangeEnd   == Maximum")]
+
+        [DataRow(  100,    0,              0,    0,    0, 0.01, DisplayName = "Minimum == RangeStart  > RangeEnd == Maximum")]
+
+        [DataRow(  -90,  -10,              0,    0,    0, 0.01, DisplayName = "Minimum  > RangeStart  < RangeEnd  < Maximum")]
+        [DataRow(  -10,  -90,              0,    0,    0, 0.01, DisplayName = "Minimum  > RangeStart  > RangeEnd  < Maximum")]
+        [DataRow(  -50,  -50,              0,    0,    0, 0.01, DisplayName = "Minimum  > RangeStart == RangeEnd  < Maximum")]
+
+        [DataRow(  110,  190,              0, 0.01, 0.01, 0.01, DisplayName = "Minimum  < RangeStart  < RangeEnd  > Maximum")]
+        [DataRow(  190,  110,              0, 0.01, 0.01, 0.01, DisplayName = "Minimum  < RangeStart  > RangeEnd  > Maximum")]
+        [DataRow(  150,  150,              0, 0.01, 0.01, 0.01, DisplayName = "Minimum  < RangeStart == RangeEnd  > Maximum")]
```
Could extract the third and sixth column for as those are always the same.

RosarioPulella commented:
In dec4699, I removed the `DisplayNames` from the `DataRows` as I felt it would not be maintained consistently and did not provide much additional info. If anyone liked it let me know so we can discuss.

RosarioPulella commented:
@michael-hawker @azchohfi and @chingucoding mind looking at this, I would like to get this merged so we can move forward with discussing, addressing issues and adjusting behavior and around the `RangeSelector`
